### PR TITLE
ART-9021 Use official release schedule API and remove auto mode for canonical_builders_from_upstream

### PR DESCRIPTION
**Pull request by Ximinhan**
Our tool is using Release_schedule lib to get release schedule through gitlab release schedule [repo](https://gitlab.cee.redhat.com/ocp-release-schedule/schedule/-/blob/master/schedules/4.15.yaml?ref_type=heads), while the repo is updated by us ref the official release schedule on pp.engineering.redhat.com.

This pr replace the function to directly get feature freeze release date from release schedule API, also add function to get GA date through API.

And update the canonical_builders_from_upstream supported mode to limit to [ True, False ] 
this change also need
https://github.com/openshift-eng/ocp-build-data/pull/4450
https://github.com/openshift-eng/ocp-build-data/pull/4451
https://github.com/openshift-eng/ocp-build-data/pull/4452
https://github.com/openshift-eng/ocp-build-data/pull/4453
https://github.com/openshift-eng/ocp-build-data/pull/4454
https://github.com/openshift-eng/ocp-build-data-validator/pull/133

### CODE REVIEW

**Comment by joepvd** [comment]
/hold

While I think the current situation is not good, I do not trust pp enough yet to have it change base images and golang builders without us being aware. 

Thinking of this approach:
- we get rid of `auto` in `canonical_builders_from_upstream`
- we get a notification if pp says branch cut happened, and we have it still on `true`

File: artcommon/artcommonlib/util.py [hunk 1]
```diff
@@ -105,6 +105,28 @@ def is_future_release_date(date_str):
...
+        release_date = datetime.strptime(release_schedules.json()[0]['date_finish'], "%Y-%m-%d")
+    return release_date
+
+
+def get_ga_release_date(major, minor):
+    """
+    Get ga release release date from release schedule API
+    """
+    release_date = None
+    release_schedules = requests.get(f'{RELEASE_SCHEDULES}/openshift-{major}.{minor}/schedule-tasks/?name=OpenShift%20Container%20Platform%20GA%20Release%20Schedule', headers={'Accept': 'application/json'})
```
Comment (locriandev) [inline artcommon/artcommonlib/util.py:124]:
```suggestion
    release_schedules = requests.get(
        f'{RELEASE_SCHEDULES}/openshift-{major}.{minor}/schedule-tasks/?name=OpenShift%20Container%20Platform%20GA%20Release%20Schedule',
        headers={'Accept': 'application/json'}
    )
```

File: doozer/doozerlib/distgit.py [hunk 2]
```diff
@@ -1614,7 +1613,7 @@ def _canonical_builders_enabled(self) -> bool:
         elif canonical_builders_from_upstream == 'auto':
             # canonical_builders_from_upstream set to 'auto': rebase according to release schedule
             try:
-                feature_freeze_date = ReleaseSchedule(self.runtime).get_ff_date()
+                feature_freeze_date = get_feature_freeze_release_date(self.runtime.group_config.vars['MAJOR'], self.runtime.group_config.vars['MINOR'])
                 return datetime.now() < feature_freeze_date
             except ChildProcessError:
```
Comment (locriandev) [inline doozer/doozerlib/distgit.py:1618]: This exception no longer needs to be caught

**Review by locriandev — CHANGES_REQUESTED** [review]

File: artcommon/artcommonlib/util.py [hunk 3]
```diff
@@ -105,6 +105,28 @@ def is_future_release_date(date_str):
         return False
 
 
+def get_feature_freeze_release_date(major, minor):
+    """
+    Get feature freeze release release date from release schedule API
+    """
+    release_date = None
+    release_schedules = requests.get(f'{RELEASE_SCHEDULES}/openshift-{major}.{minor}/schedule-tasks/?name=Feature%20Development%20for%20{major}.{minor}', headers={'Accept': 'application/json'})
```
[the remark this answers is not in the document]
Comment (Ximinhan) [reply artcommon/artcommonlib/util.py]: the server is valid enough, I think the check response code is should be catch the invalid response, and when its empty the return valid is None and distgit check will raise the error when compare None and datetime.now()

File: artcommon/artcommonlib/util.py [hunk 4]
```diff
@@ -105,6 +105,28 @@ def is_future_release_date(date_str):
         return False
 
 
+def get_feature_freeze_release_date(major, minor):
```
[the remark this answers is not in the document]
Comment (Ximinhan) [reply artcommon/artcommonlib/util.py]: ok to move

**Comment by Ximinhan** [comment]
> /hold
> 
> While I think the current situation is not good, I do not trust pp enough yet to have it change base images and golang builders without us being aware.
> 
> Thinking of this approach:
> 
> * we get rid of `auto` in `canonical_builders_from_upstream`
> * we get a notification if pp says branch cut happened, and we have it still on `true`

I don't get what your approach means.
Do you need some reassurance about the pp.engineering.redhat.com? Instead, I think the release schedule repo may lack of update when  pp.engineering.redhat.com has a new change.

File: artcommon/artcommonlib/util.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Ximinhan) [reply artcommon/artcommonlib/util.py:138]: actually when I tried to move to release_util find there is a chain reference, get_inflight use get_assembly_release_date and get_ocp_version_from_group and is_future_release_date, I don't see a big difference to split into two util libs, instead we the two utils may have cross reference. So I would keep it in util.py

File: artcommon/artcommonlib/util.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Ximinhan) [reply artcommon/artcommonlib/util.py]: retry added

**Pull request closed without merging**